### DISCUSSION TITLE
Callback for USBDevice.mounted() or similar

### DISCUSSION THREAD

**Discussion by timonsku** [hathach/tinyusb#283] [Ideas]
I was wondering if there is a callback for USBDevice.mounted() or something equivalent that informs me when the device was connected to or disconnected from the host.
I'm writing a library that depends on checking this periodically, would be nice if I can do this without modifying the core. I want to avoid the user having to manually call an update function of the library.

I sifted through the code a bit but couldn't find something obvious.

**Comment by hathach** [comment]
@PTS93 there is an callback for  tud_mount_cb/tud_umount_cb in the stack.
https://github.com/hathach/tinyusb/blob/master/src/device/usbd.h#L97
cdc_msc example use it to set the blinking interval to inform the connection status
https://github.com/hathach/tinyusb/blob/master/examples/device/cdc_msc/src/main.c#L78

Are you looking for an Arduino callback function ? You can just use the callback directly, though, as convention, Arduino callback should be passed to USBDevice with something like setMountedCallback() or something.

Sorry for the late response, hope this answer your question.

**Comment by timonsku** [comment]
Yes ultimately I will need it in Arduino. Where would you suggest that I add that setMountedCallback() wrapper for tud_mount_cb/tud_umount_cb? The arduino library itself does not seem to implement any global USB functions but only the specific USB classes. Are those things being done in the Arduino core side?

**Comment by timonsku** [comment]
Ah I think I finally understood the setup. So there are 2 Arduino libs essentially. One setting up the general USB and the other is for implementing specific devices like HID or MSD?
So I assume here would be a good place for this callback wrapper?
https://github.com/adafruit/Adafruit_TinyUSB_ArduinoCore/blob/09968245ad675377f2265717f0e138d5dccb39a3/Adafruit_USBD_Device.h

**Comment by hathach** [comment]
@PTS93 right, the core lib implement device + CDC, since cdc is required for Arduino core to perform DFU touch1200 and Serial. The MSC + HID and other higher layers are implemented as wrapper for higher class, which is subject to be changed more frequent. It is kind of complicated, but I don't any better way. 

an setMountedCallback() with an function pointer should be fine, you should also add the setUmountedCallback() as well.

**Comment by hathach** [comment]
API is implemented https://github.com/adafruit/Adafruit_TinyUSB_ArduinoCore/blob/master/Adafruit_USBD_Device.h#L75

**Comment by timonsku** [comment]
but that is not a callback or did you link to the wrong place?
Sorry I meant to integrate this but got side tracked quite a bit.

**Comment by hathach** [comment]
right, I forgot what I have said, and thought it is merely for polling, will add one callback later on
